packages/driver/src/cypress/runner.ts

```diff
@@ -492,7 +493,78 @@ const hasOnly = (suite) => {
   )
 }
 
-const normalizeAll = (suite, initialTests = {}, setTestsById, setTests, getRunnableId, getHookId, getOnlyTestId, getOnlySuiteId, createEmptyOnlyTest) => {
+// Removes a suite and any of it's hooks/tests. Also removes the reference to itself so that it can be GC'd
+const pruneSuite = (emptySuite) => {
+  emptySuite.cleanReferences()
+
+  if (emptySuite.parent) {
+    const parentSuites = emptySuite.parent.suites.filter((suite) => suite !== emptySuite)
+
+    emptySuite.parent.suites = parentSuites
+    emptySuite.parent._onlySuites = emptySuite.parent._onlySuites.filter((suite) => parentSuites.includes(suite))
+    emptySuite.parent = null
+  }
+}
+
+// When in open mode and a "testFilter" is active, tests/suites that do not match the test filter
+// need to be removed as they might have modifiers (.only) that would affect the matched set of tests.
+// This function will recursively iterate through all of the suites and filter out any tests that do not
+// match the specified filter. If the suite is empty after removing the tests, the suite is also removed.
+const pruneEmptySuites = (rootSuite, testFilter: NonNullable<TestFilter>) => {
+  // We want to start at the lowest level so recurse first. We want to prune child suites before parents
+
+  let totalUnfilteredTests = 0
+
+  for (const suite of [...rootSuite.suites]) {
+    totalUnfilteredTests += pruneEmptySuites(suite, testFilter)
+  }
+
+  if (!rootSuite.suites.length && !rootSuite.tests.length) {
+    pruneSuite(rootSuite)
+  }
+
+  if (rootSuite.tests.length) {
+    totalUnfilteredTests += rootSuite.tests.length
+
+    const tests: any[] = []
+    const onlyTests: any[] = []
+
+    for (const test of rootSuite.tests) {
+      // Tests/suites of the shape "it('should', { browser: 'chrome' }, ...)" will have their title
+      // updated with a skipped message. Even if the test is skipped we should still show it in the reporter
+      // so we match the "fullTitle" with the skipped message removed.
+      const fullTitle = test.fullTitle().replaceAll(SKIPPED_DUE_TO_BROWSER_MESSAGE, '')
+
+      if (testFilter.includes(fullTitle)) {
+        tests.push(test)
+
+        if (rootSuite._onlyTests.includes(test)) {
+          onlyTests.push(test)
+        }
+      } else {
+        delete test.fn
+      }
+    }
+
+    rootSuite.tests = tests
+    rootSuite._onlyTests = onlyTests
+
+    if (!rootSuite.tests.length && !rootSuite.suites.length) {
+      pruneSuite(rootSuite)
+    }
+  }
+
+  return totalUnfilteredTests
+}
+
+const normalizeAll = (suite, initialTests = {}, testFilter, setTestsById, setTests, getRunnableId, getHookId, getOnlyTestId, getOnlySuiteId, createEmptyOnlyTest) => {
+  let totalUnfilteredTests = 0
```

Comment:
should this default value be the total number of tests in the spec, not zero?

Reply:
Thought looking at the UI, is this meant to represent the totalFilteredTests?

Reply:
I believe Zach structured it this way to simplify accounting for filtered nested runnables (eg `.only` and "skipped due to browser" conditions). Was easier to sum up matching elements than it was to deduct non-matches

The variable naming appears correct to me - it's used for the "total" value in the output of `Header.tsx` (`x/y tests`)